File: src/main/java/io/scriptor/util/ChannelInputStream.java
package io.scriptor.util;

import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;

public class ChannelInputStream extends ExtendedInputStream {

    private final FileInputStream stream;
    private final FileChannel channel;

    public ChannelInputStream(final @NotNull String name) throws FileNotFoundException {
        super();
        this.stream = new FileInputStream(name);
        this.channel = stream.getChannel();
    }

    public ChannelInputStream(final @NotNull File file) throws FileNotFoundException {
        super();
        this.stream = new FileInputStream(file);
        this.channel = stream.getChannel();
    }

    public ChannelInputStream(final @NotNull FileDescriptor fd) {
        super();
        this.stream = new FileInputStream(fd);
        this.channel = stream.getChannel();
    }

    @Override
    public int read() throws IOException {
        return stream.read();
    }

    @Override
    public int read(final byte @NotNull [] b) throws IOException {
        return stream.read(b);
    }

    @Override
    public int read(final byte @NotNull [] b, final int off, final int len) throws IOException {
        return stream.read(b, off, len);
    }

    @Override
    public byte @NotNull [] readAllBytes() throws IOException {
        return stream.readAllBytes();
    }

    @Override
    public byte @NotNull [] readNBytes(final int len) throws IOException {
        return stream.readNBytes(len);
    }

    @Override
    public int readNBytes(final byte @NotNull [] b, final int off, final int len) throws IOException {
        return stream.readNBytes(b, off, len);
    }

    @Override
    public long skip(final long n) throws IOException {
        return stream.skip(n);
    }

    @Override
    public void skipNBytes(final long n) throws IOException {
        stream.skipNBytes(n);
    }

    @Override
    public int available() throws IOException {
        return stream.available();
    }

    @Override
    public void mark(final int readlimit) {
        stream.mark(readlimit);
    }

    @Override
    public void reset() throws IOException {
        stream.reset();
    }

    @Override
    public boolean markSupported() {
        return stream.markSupported();
    }

    @Override
    public long transferTo(final @NotNull OutputStream out) throws IOException {
        return stream.transferTo(out);
    }

    @Override
    public long tell() throws IOException {
        return channel.position();
    }

    @Override
    public void seek(final long pos) throws IOException {
        channel.position(pos);
    }

    @Override
    public long size() throws IOException {
        return channel.size();
    }

    @Override
    public void read(final @NotNull ByteBuffer buffer) throws IOException {
        channel.read(buffer);
    }

    @Override
    public void close() throws IOException {
        stream.close();
    }
}

File: src/main/java/io/scriptor/util/ChannelInputStream.kt
package io.scriptor.util

import java.io.File
import java.io.FileDescriptor
import java.io.FileInputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.channels.FileChannel

class ChannelInputStream : ExtendedInputStream {

    private val stream: FileInputStream
    private val channel: FileChannel

    constructor(name: String) : super() {
        this.stream = FileInputStream(name)
        this.channel = stream.channel
    }

    constructor(file: File) : super() {
        this.stream = FileInputStream(file)
        this.channel = stream.channel
    }

    constructor(fd: FileDescriptor) : super() {
        this.stream = FileInputStream(fd)
        this.channel = stream.channel
    }

    override fun read(): Int {
        return stream.read()
    }

    override fun read(b: ByteArray): Int {
        return stream.read(b)
    }

    override fun read(b: ByteArray, off: Int, len: Int): Int {
        return stream.read(b, off, len)
    }

    override fun readAllBytes(): ByteArray {
        return stream.readAllBytes()
    }

    override fun readNBytes(len: Int): ByteArray {
        return stream.readNBytes(len)
    }

    override fun readNBytes(b: ByteArray, off: Int, len: Int): Int {
        return stream.readNBytes(b, off, len)
    }

    override fun skip(n: Long): Long {
        return stream.skip(n)
    }

    override fun skipNBytes(n: Long) {
        stream.skipNBytes(n)
    }

    override fun available(): Int {
        return stream.available()
    }

    override fun mark(readlimit: Int) {
        stream.mark(readlimit)
    }

    override fun reset() {
        stream.reset()
    }

    override fun markSupported(): Boolean {
        return stream.markSupported()
    }

    override fun transferTo(out: OutputStream): Long {
        return stream.transferTo(out)
    }

    override fun tell(): Long {
        return channel.position()
    }

    override fun seek(pos: Long) {
        channel.position(pos)
    }

    override fun size(): Long {
        return channel.size()
    }

    override fun read(buffer: ByteBuffer) {
        channel.read(buffer)
    }

    override fun close() {
        stream.close()
    }
}
